### PR TITLE
Safeguard round function decimal places overflow

### LINKED ISSUE #6133: panic in round() math function
### Describe the bug

I noticed this bug when reviewing panics for this issue https://github.com/apache/arrow-datafusion/issues/3316
https://github.com/apache/arrow-datafusion/blob/a38480951f40abce7ee2d5919251a1d1607f1dee/datafusion/physical-expr/src/math_expressions.rs#L215
For `round(num, decimal_place)` function (`dicimal_place` has `Int64` signature), when `decimal_place` overflow `Int32` range, it will panic.

### To Reproduce

```sql
-- max Int32 is 2147483647
select round(3.14, 2147483648);
```

```
DataFusion CLI v23.0.0
❯ select round(3.14, 2147483648);

thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: TryFromIntError(())', /Users/alamb/Software/arrow-datafusion2/datafusion/physical-expr/src/math_expressions.rs:238:74
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

### Expected behavior

An error saying that `2147483648` is out of range

### Additional context

For a quick fix, the code can return an Execution Error instead of panicking. 

However, it seems better to return a Planning Error during the Parsed Tree -> Logical Plan (aka binding) stage: (1) the `round(num, decimal_place)` implementation relies on the standard library, which only accepts `i32` for `decimal_place`. As a result, the function signature for decimal_place should be `int32`. (2) be more compatible with Postgres, it will return planning error for such query and the function signature for `decimal_place` is 32 bit int.

Currently, all integers are parsed as i64 during planning. Implementing the suggested change might require more modifications. I would like to know if such changes are desired.

### PR DESCRIPTION
**Pull request by tz70s**
# Which issue does this PR close?

Closes #6133.

# Rationale for this change

Add runtime check rather than crash the system.

# What changes are included in this PR?

Safeguard the panic from round function decimal places overflow.

In #6133 stated there could be a type check in binding phase, only add the runtime check in this PR though I do agree having error throwing out in the logical planning might be better.

# Are these changes tested?

Yes

# Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/physical-expr/src/math_expressions.rs [hunk 1]
```diff
@@ -226,21 +227,23 @@ pub fn round(args: &[ArrayRef]) -> Result<ArrayRef> {
                     }
                 )) as ArrayRef)
             }
-            ColumnarValue::Array(decimal_places) => Ok(Arc::new(make_function_inputs2!(
```
Comment (tz70s) [inline datafusion/physical-expr/src/math_expressions.rs:229]: I basically give up using the existing macro (`make_function_inputs2`), not sure is it worth changing it to be possibly return `Result` within the `$FUNC`.

**Comment by tz70s** [comment]
Maybe we can simply change the `Round` to accept i32 here: https://github.com/apache/arrow-datafusion/blob/main/datafusion/expr/src/function.rs#L651-L654

**Comment by tz70s** [comment]
> Maybe we can simply change the `Round` to accept i32 here: https://github.com/apache/arrow-datafusion/blob/main/datafusion/expr/src/function.rs#L651-L654

Postgres output:

```
postgres=# select round(5.4323, 2147483648);
ERROR:  function round(numeric, bigint) does not exist
LINE 1: select round(5.4323, 2147483648);
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
```

File: datafusion/physical-expr/src/math_expressions.rs (hunk 1 above)
Comment (2010YOUY01) [reply datafusion/physical-expr/src/math_expressions.rs:229]: Nice work! This overflow issue can be solved during binding in the future(hopefully...), I think it's not necessary to change the existing macro 🤔

File: datafusion/physical-expr/src/math_expressions.rs [hunk 2]
```diff
@@ -290,6 +295,14 @@ pub fn round(args: &[ArrayRef]) -> Result<ArrayRef> {
     }
 }
 
+fn round_downcast_decimal_places(decimal_place: i64) -> Result<i32> {
```
Comment (2010YOUY01) [inline datafusion/physical-expr/src/math_expressions.rs:298]: I noticed most small helpers inside execution are macros to guarantee inline for performance, should we make this function macro?

Comment (tz70s) [reply datafusion/physical-expr/src/math_expressions.rs]: Thanks, I made it as a macro instead!

Comment (viirya) [reply datafusion/physical-expr/src/math_expressions.rs]: You could inline a function too. Not necessary to be macros actually as macros are more difficult to read.

File: datafusion/physical-expr/src/math_expressions.rs [hunk 3]
```diff
@@ -287,6 +288,16 @@ pub fn random(args: &[ColumnarValue]) -> Result<ColumnarValue> {
...
+macro_rules! round_downcast_decimal_places {
+    ($dp:expr) => {
+        $dp.try_into().map_err(|_| {
+            DataFusionError::Execution(
+                "round function requires decimal_places as 32 bit integers".to_string(),
+            )
+        })
+    };
+}
+
```
Comment (viirya) [inline datafusion/physical-expr/src/math_expressions.rs:300]: The input/output types are always i64/i32. This isn't necessary to be a macro, I guess.

Comment (viirya) [reply datafusion/physical-expr/src/math_expressions.rs:300]: You can have a inline function instead.

**Comment by alamb** [comment]
@viirya (or @liukun4515 ) do you have time to review / approve this PR? I haven't been following closely and I think you are the experts in Decimal implementation in DataFusion

**Comment by alamb** [comment]
I wonder if https://github.com/apache/arrow-datafusion/pull/6833 is similar functionality

**Comment by viirya** [comment]
I think the change is so out-of-dated from latest branch so the diffs are unclear what are actual change from this PR. It'd be better to sync up with latest branch.

**Pull request closed without merging**